File: chess/src/gameLogic/pieces/bishop.js
import Piece from "./piece";
import Coord from "../helpers/coordinates";

export default class Bishop extends Piece {
  constructor(...args) {
    super(args);
    this.type = "B";
  }

  copy() {
    return new Bishop(...super.getInfos());
  }

  computeMoves(pieces, needToVerify) {
    this.moves.erase();
    let flags = [...Array(4)].fill(true);
    const x = this.coord.x;
    const y = this.coord.y;

    const checkLine = (flagNumber, comparison, coord) => {
      if (flags[flagNumber]) {
        if (comparison()) {
          flags[flagNumber] = false;
        } else {
          let otherPiece = pieces.findByCoord(coord);
          if (otherPiece) {
            if (otherPiece.team !== this.team) {
              this.moves.add(this, coord, "X", otherPiece);
            }
            flags[flagNumber] = false;
          } else this.moves.add(this, coord, "M", null);
        }
      }
    };

    for (let i = 1; i < 8; i++) {
      checkLine(0, () => x - i < 0 || y - i < 0, new Coord(x - i, y - i));
      checkLine(1, () => x + i > 7 || y - i < 0, new Coord(x + i, y - i));
      checkLine(2, () => x - i < 0 || y + i > 7, new Coord(x - i, y + i));
      checkLine(3, () => x + i > 7 || y + i > 7, new Coord(x + i, y + i));
    }

    if (needToVerify) super.verifyMoves(pieces);
  }
}

File: engine/GameLogic/Pieces/Bishop.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ChessEngine.GameLogic.Utils;

namespace ChessEngine.GameLogic.Pieces
{
    public sealed class Bishop : Piece
    {
        public Bishop(bool team, Coord coord)
            : base(team, coord) { }

        public override void ComputeMoves(List<Piece> pieces, bool needToVerify)
        {
            bool[] flags = { true, true, true, true };

            void Square(int flag, Coord coord)
            {
                if (flags[flag])
                {
                    if (!coord.IsValid())
                        flags[flag] = false;
                    else
                    {
                        Piece otherPiece = new Piece();
                        foreach (Piece p in pieces)
                            if (p.Coord == coord)
                                otherPiece = p;

                        if (!otherPiece.isDefault())
                        {
                            if (otherPiece.Team != Team)
                            {
                                Moves.Add(new Move('X', coord, otherPiece));
                            }
                            flags[flag] = false;
                        }
                        else Moves.Add(new Move('M', coord));
                    }

                }
            }

            for (int i = 1; i <= 8; i++)
            {
                Square(0, new Coord(Coord.x - i, Coord.y - i));
                Square(1, new Coord(Coord.x + i, Coord.y - i));
                Square(2, new Coord(Coord.x - i, Coord.y + i));
                Square(3, new Coord(Coord.x + i, Coord.y + i));
            }
            if (needToVerify) base.VerifyMoves(pieces);
        }
    }
}
